src/style-spec/function/evaluation_context.js

```diff
@@ -0,0 +1,245 @@
+import type { Type } from './types';
+import type { Value } from './values';
+import type { InterpolationType } from './definitions/curve';
+import type { Feature } from './index';
+
+const geometryTypes = ['Unknown', 'Point', 'LineString', 'Polygon'];
+const types = {
+    Null: NullType,
+    Number: NumberType,
+    String: StringType,
+    Boolean: BooleanType,
+    Color: ColorType,
+    Object: ObjectType,
+    Value: ValueType
+};
+
+class RuntimeError {
+    name: string;
+    message: string;
+    constructor(message) {
+        this.name = 'ExpressionEvaluationError';
+        this.message = message;
+    }
+
+    toJSON() {
+        return `${this.message}`;
+    }
+}
+
+// don't call this 'assert' because build/min.test.js checks for 'assert('
+// in the bundled code to verify that unassertify is working.
+function ensure(condition: any, message: string) {
+    if (!condition) throw new RuntimeError(message);
+    return true;
+}
+
+
+module.exports = () => ({
+    types: types,
+
+    ensure: ensure,
+    error: (msg: string) => ensure(false, msg),
+
+    at: function (index: number, array: Array<Value>) {
+        ensure(index >= 0 && index < array.length,
+            `Array index out of bounds: ${index} > ${array.length}.`);
+        ensure(index === Math.floor(index),
+            `Array index must be an integer, but found ${String(index)} instead.`);
+        return array[index];
+    },
+
+    get: function (obj: {[string]: Value}, key: string, name?: string) {
+        ensure(this.has(obj, key, name), `Property '${key}' not found in ${name || `object`}`);
+        return obj[key];
+    },
+
+    has: function (obj: {[string]: Value}, key: string, name?: string) {
+        ensure(obj, `Cannot get property ${key} from null object${name ? ` ${name}` : ''}.`);
+        ensure(typeof obj === 'object', `Expected ${name || 'value'} to be of type Object, but found ${toString(typeOf(obj))} instead.`);
+        return typeof obj[key] !== 'undefined';
+    },
+
+    contains: function (value: Value, array: Array<Value>) {
+        const type = typeOf(value).kind;
+        ensure(type !== 'Object' && type !== 'Array' && type !== 'Color',
+            `"contains" does not support values of type ${type}`);
+        return array.indexOf(value) >= 0;
+    },
+
+    typeOf: function (x: Value): string {
+        return toString(typeOf(x));
+    },
+
+    as: function (value: Value, expectedType: Type, name?: string) {
+        assert(isValue(value), `Invalid value ${JSON.stringify(value)}`);
+        assert(expectedType.kind, `Invalid type ${JSON.stringify(expectedType)}`);
+
+        let type;
+        let typeError = false;
+        if (expectedType.kind === 'Null') {
+            typeError = value === null;
+        } else if (expectedType.kind === 'Value') {
+            typeError = true;
+        } else if (expectedType.kind !== 'Array') {
+            typeError = typeof value !== expectedType.kind.toLowerCase();
```

Comment:
nitpick: `toLowerCase` will be called on almost every `get`; might be preferable to make a lookup table for kinds instead